Fixes for rendering text with bottom/right alignment

discnl opened the pull request:
Fixes several generic and Windows specific drawing issues, almost all with text, that can be seen by using `DrawItemText()` with bottom and right alignment in the render sample starting with 8388a70.

Some notes:

Both commits that get reverted were originally a result of `wxDC::DrawLabel()`'s off-by-one with right aligned text (see also 129cf3a). The off-by-ones seem to have always been there and it's likely that it affects other user/wx code as well in silent ways. So far I have only noticed small offset differences that got more noticeable and required changing `wxGCDCImpl::DoGetTextExtent()` (see d9d0ff9).

The changes to the dataview sample included in this PR also help with demonstrating some of the problems described in commit messages. I tried to make the changes a bit generic and not too debug specific. For the render sample they are more crude and could perhaps use similar alignment options (it already has them for horizontal alignment), so that a single `DrawItemSelectionRect()` can be used to also show e.g. bottom alignment.

Finally regarding the new usage of `::GetThemeTextExtent()`: Here it's only used for positioning multi-line text but in general it results in better calculated extents for themed text. It should eventually probably be refactored to be used somehow by wx' text measurement code elsewhere.

discnl committed:
Add wxDataViewCtrl debug option to display render bounds

For debugging convenience define DEBUG_RENDER_EXTENTS to draw a red
rectangle around a custom rendered cell's full rectangle, and a green
rect for the extent of the item appearing inside it. Custom renderers
ordinarily should not draw outside of the green rect. A notable
exception is drawn text, particularly multi-line ones.

discnl committed:
Add render-related options to dataview sample

To more easily expose problems add options to the dataview sample
related to rendering of items (applying mostly to the MyListModel page
only):

* Use left/centre/right alignment (Ctrl+1/2/3)
* Use top/centre/bottom alignment (Ctrl+4/5/6)
* Toggle tall row usage (Ctrl+7)
* Toggle keep on using small wx logo, regardless of row size (Ctrl+8)
* Toggle multi-line text usage (Ctrl+9)

discnl committed:
Avoid assert failures in render sample with generic renderer

Some generic render functions aren't implemented and result in obtrusive
assert failures when switching to use the generic renderer in the render
sample. Instead show text saying the drawing function is not implemented
in the location where the sample would ordinarily draw the control.

discnl committed:
Demonstrate wxRenderer::DrawItemText() alignment in render sample

This is intended for wxMSW to show differences between themed text
drawing (with wxRendererXP::DrawItemText() under Vista+) and
wxRendererGeneric::DrawItemText(). Use existing option to switch between
generic and native renderer (Ctrl-G) in the render sample to observe
the differences.

discnl committed:
Fix centered vertical and bottom aligned themed text drawing

When using non-TOP vertical alignment with wxRendererXP::DrawItemText()
either DT_VCENTER or DT_BOTTOM gets passed to DrawThemeTextEx() but
without the DT_SINGLELINE flag which is required for those alignment
flags to take effect, resulting in text always being top aligned.

Fix by passing DT_SINGLELINE when using either alignment, but only for
single lines as multi-lines are rendered as a single line with invisible
newline character. Draw multi-line text using top alignment and deal
with vertical alignment ourselves, using GetThemeTextExtent() to get
an accurate extent of the text.

discnl committed:
Fix off-by-ones with bottom and right aligned positioning

wxDC::DrawLabel() positions the text one to the left with wxALIGN_RIGHT
alignment and one up with wxALIGN_BOTTOM alignment. The same occurs in
wxControlRenderer::DrawBitmap() for bitmaps. Both off-by-ones exist
since inception in respectively 4d3c4c2f94 and bc60c3d699.

This fix vertically aligns wxALIGN_BOTTOM drawing of wxDC::DrawLabel()
with themed text drawing through wxRendererXP::DrawItemText().

discnl committed:
Undo mimicking faulty wxDC::DrawLabel() wxALIGN_RIGHT positioning

Revert b642747fd2 (Fix right aligned text position in wxDVC on MSW with
system theme, 2015-09-30) which in wxRendererXP::DrawItemText() mimicks
the off-by-one with right alignment of wxDC::DrawLabel() that got fixed
by the previous commit.

There's not a similar change copying wxDC::DrawLabel()'s former
off-by-one behaviour with wxALIGN_BOTTOM that would need reverting.

Both alignments now result in the same text positioning with both
drawing functions.

discnl committed:
Undo removal of text alignment for wxDataViewCtrl's RenderText()

Revert e2e7d3d391 (Fix wxELLIPSIZE_END with wxALIGN_RIGHT in wxMSW
wxDataViewCtrl, 2016-03-18). This fix is no longer needed after
the previous commit reverted b642747fd2.

Reverting also allows for text to be drawn vertically aligned again
instead of always using top-left alignment. While the difference in
appearance by not having alignment can be minor with default row
heights, it becomes more noticeable with taller rows:
a wxDataViewCheckIconText column with a tall icon will have its text
stuck to the top of a row while other columns have their text vertically
centered. This already occurs by default when not explicitly specifying
an alignment (wxDVR_DEFAULT_ALIGNMENT) which results in
wxALIGN_CENTRE_VERTICAL being used for row alignment when rendering.

discnl committed:
Improve text extent rounding in wxGCDCImpl::DoGetTextExtent()

Change wxGCDCImpl::DoGetTextExtent() from rounding to the nearest
integer to rounding up: if e.g. height is 15.3 then 16 pixels should be
used for height, and not 15. Rounding was previously improved from
casting (which appears to be the initial code) in 0417955ddb (adding
correct filling area to arc, correct rounding and clipping, 2007-10-21).

Issues with nearest rounding became more visible after off-by-one fixes
for wxDC::DrawLabel() with wxALIGN_RIGHT and wxALIGN_BOTTOM positioning.
When using e.g. wxALIGN_RIGHT with a width of 72.4 the text could now be
drawn beyond the extent on the right because the text is not offset by
one to the left any longer.

vadz commented:
This finally looks like the real solution, thanks a lot for finding it!

Do I understand correctly that this also fixes [#18487](https://trac.wxwidgets.org/ticket/18487)?

In any case, I think this should be merged but let's wait for a couple of days to see if other people who have worked on this code (@a-wi, @TcT2k) have any comments.

Thanks again!

discnl committed:
Work around text extent differences resulting in clipped text

Under at least some versions of Windows 10 with wxDVC themed text can be
clipped horizontally because of discrepancies between the text extent
as drawn by DrawThemeTextEx() and calculated with GetTextExtent()
earlier.

Work around the issue by always trying to use GetThemeTextExtent() in
DrawItemText(), not just for alignment of multi-line strings, and adjust
for any width differences similar to the existing adjustment for height.

See #18487.

discnl commented:
> Do I understand correctly that this also fixes [#18487](https://trac.wxwidgets.org/ticket/18487)?

No, that requires width adjustments while so far this was conservative and only uses `::GetThemeTextExtent()` when it's needed for multi-lines with non-top alignment and then only does height adjustments. I went ahead and made a workaround for #18487 in e19f903.

discnl commented on src/msw/renderer.cpp:
```diff
@@ -1093,12 +1099,86 @@ void wxRendererXP::DrawItemText(wxWindow* win,
+            if ( align & wxALIGN_BOTTOM )
+                textFlags |= DT_BOTTOM | DT_SINGLELINE;
+            else if ( align & wxALIGN_CENTER_VERTICAL )
+                textFlags |= DT_VCENTER | DT_SINGLELINE;
```
Because Windows XP doesn't have `GetThemeTextExtent()` this code gets used and while the addition of `DT_SINGLELINE` results in bottom and centered vertical aligned text to work it also renders multi-lines as a single line now. Formerly all text was always top aligned vertically whcih does render multi-lines correctly. I think here it should either:

a) (preferred) Use `DT_TOP` (without `DT_SINGLELINE`) if non-top vertical alignment is used and text contains a newline. I don't like the cost of checking the latter (which isn't done any longer with e19f903) but at least multi-lines then work as before: always top aligned but otherwise rendered correctly and of course single-line non-top aligned text gets properly aligned now because of using `DT_SINGLELINE`.

or:

b) Don't use `DT_SINGLELINE` at all like before, restoring previous behaviour under XP only: all text is top aligned.

vadz replied:
I'm fine with either (a) or (b). I'd like to say we don't need to care about XP at all any longer, but there are probably still people using it...

discnl replied:
Sorry this can't actually apply to XP as it doesn't have `DrawThemeTextEx()`, so it's only regarding what to do if `GetThemeTextExtent()` fails which I haven't seen happening yet so I'll keep things as-is right now and maybe change it after the code settles down more.

discnl commented:
The later added e19f903 completely breaks (themed) ellipsization because it adjusts the rect width to fit the text exactly. What I think it unfortunately should do instead is check if there's a difference in text size between `dc.GetTextExtent()` (I haven't tried with `dc.GetMultiLineTextExtent()` so far) and `GetThemeTextExtent()` and adjust the rect only by that. I don't like it much since there's rarely a difference and because of possible other side-effects but it does fix [#18487](https://trac.wxwidgets.org/ticket/18487). So unless there's another suggestion I'll be going for something like that.

discnl committed:
Correct slightly misleading comment

Comment in wxRendererXP::DrawItemText() could be interpreted as
DrawThemeTextEx() being available for some XP editions but it is Vista+
only, so correct it.

discnl commented:
If the sample changes are OK then this is mostly done I think, though I'd like to mention the impact this can have on [not respecting](https://github.com/wxWidgets/wxWidgets/pull/2446/commits/a8e5284d6cf72e9b8c7bee80cff52b8e9d2cc36e#diff-bf73bfcbc431b2b67aca5ce1cfa8ca687e5d1ec684e83daee5225ec59a43dcefR1188-R1190) the drawing rect limits of `wxRendererXP::DrawItemText()` any longer. I'm inclined to confine the height changes to the [necessary](https://github.com/wxWidgets/wxWidgets/pull/2446/commits/68e79294cb11a6b0ec3b8aa3fc5304a46244a6cf#diff-bf73bfcbc431b2b67aca5ce1cfa8ca687e5d1ec684e83daee5225ec59a43dcefR1110-R1113) cases only again. That would still leave height adjustments to occur for multi-line text that's non-top aligned, and because with `wxDVC` centered is the default vertical alignment it changes the most common case.

Unlike the width difference with `GetThemeTextExtent()`,which is usually non-existing or minor, the height difference with multi-line text is larger because `wxDVC` makes use of `wxWindow::GetTextExtent()` which for wxMSW eventually uses `GetTextExtentPoint32()` which treats text as a single line (`DT_CALCRECT` is already used in a few cases in wx to get the actual text extent). The [docs](https://github.com/wxWidgets/wxWidgets/blob/ebec1ff9f608f9547a8fe0ce9b07253b3e440551/interface/wx/window.h#L2139-L2140) mention:

> Gets the dimensions of the string as it would be drawn on the window with the currently selected font.

which is a bit open to interpretation, if said window drawer only is supposed to work for single lines, but I would (always) expect it to work for multi-line text based on that description. Out of three only wxGTK seems to behave that way though.

discnl commented on src/generic/dcpsg.cpp:
```diff
@@ -2264,16 +2264,16 @@ void wxPostScriptDCImpl::DoGetTextExtent(const wxString& string,
     /* return size values */
     if ( x )
-        *x = int(sum * fontSize) / 1000;
+        *x = (wxCoord) ceil(sum * fontSize / 1000.0f);
```
Besides the added `ceil()` the rounding here is different but it restores the calculation to how it was before the bottom changes of [dcpsg.cpp](https://github.com/wxWidgets/wxWidgets/commit/b5a554b9a64ebbeeae7c7d68128f705e0ee3d4e2#diff-568c01f02e655c348ab79f317d781a7b48554884483b46ed7d5723294f4340d4), and doesn't convert to int early. So this should be OK again as the early conversion didn't seem intentional.

paulcor replied:
The conversion to `int` was intentional, and I assert that it did not change the result of the calculation.

discnl replied:
Thanks for clarifying! It did assert here but not for what I would say are practical values of `sum`. What I was worried about when originally seeing the cast to `int` without commit context was that it is required and my change would break some fix. Because prior to b5a554b9a64ebbeeae7c7d68128f705e0ee3d4e2 (which helpfully didn't indicate it being for a render fix) the code was also similar to my changes (sans `ceil()`) it seemed safe but figured I would mention it anyway in case someone else noticed the same.

discnl commented:
In case this will be merged later: let me know so I can rebase and also deal with the autosquash conflicts.

vadz commented:
Thanks again!

I didn't test these changes, but they look correct to me and explanations in the commit messages and comments are convincing, so, assuming they work for you, I'd like to merge it right now. Should I do it after squashing the fixup commits or do you prefer to do it yourself?

One thing I wonder about is whether we should add `wxCeil()` similar to `wxRound()`, i.e. returning an int rather than double. We could remove a number of casts to `wxCoord` then... but this is a purely cosmetic issue, of course.

discnl commented:
> I didn't test these changes, but they look correct to me and explanations in the commit messages and comments are convincing, so, assuming they work for you,

Not convinced by the sample changes but I prefer possible improvements later over reverting.

> I'd like to merge it right now. Should I do it after squashing the fixup commits or do you prefer to do it yourself?

I think I should do them (later today) as there will be conflicts and I would like to keep the related width commit separate from the height one.

> One thing I wonder about is whether we should add `wxCeil()` similar to `wxRound()`, i.e. returning an int rather than double. We could remove a number of casts to `wxCoord` then... but this is a purely cosmetic issue, of course.

It could also help with knowing at a glance which parts of the code have been reviewed (or well, changed at a minimal...) more recently. I wonder about possible confusion with `ceil()` though, perhaps something more domain specific such as `wxRoundToPixels()` (provided these are even pixels...). Or, since I think there's a need for `floor()` as well, e.g. `wxFloorPixels()` and `wxCeilPixels()`.

discnl commented:
Rebased and squashed. Two code changes:

* Removed fe66d52fb4e51ee675b2465b36ef03c9d5a27420. For some DC's it seems a good idea to have similar rounding changes to at least their `GetCharWidth()`, and I think it's better to group (and more properly test) such changes for another PR.

* [Confined](https://github.com/wxWidgets/wxWidgets/pull/2446/commits/90ba137f2059355f9dd85de932653dd6a4e317e6#diff-bf73bfcbc431b2b67aca5ce1cfa8ca687e5d1ec684e83daee5225ec59a43dcefR1193-R1198) drawing rect height changes to multi-line non-top vertical aligned text only again. While always adjusting for any height difference also fixed some alignment issue with single lines, that issue then remains for non-themed drawing. Again, better to solve it separately.

paulcor commented:
> One thing I wonder about is whether we should add `wxCeil()` similar to `wxRound()`

`wxRound()` exists not to avoid a cast, but because there is no `round()` in C++03. It does have `ceil()` and `floor()`, so there is no compelling reason to have a `wxCeil()` or `wxFloor()`. Casting to `int` just isn't that hard.

discnl commented:
> > One thing I wonder about is whether we should add `wxCeil()` similar to `wxRound()`
> 
> `wxRound()` exists not to avoid a cast, but because there is no `round()` in C++03. It does have `ceil()` and `floor()`, so there is no compelling reason to have a `wxCeil()` or `wxFloor()`. Casting to `int` just isn't that hard.

Indeed, no conditional code is needed. Even `ceilf()`, `ceill()` and floor equivalents appear readily available (but would be completely unused with current code because only arguments with type double are used both with `ceil()` and `floor()` calls). What's left is a `wxCeil()`/`wxFloor()` asserts checking for limits, similar to `wxRound()`, but that doesn't seem very valuable to me. All in all I think they would then just mostly pollute wx/math.h for the sake of not casting on about a dozen lines currently.

vadz commented:
Thanks again for your work on this, I'm going to merge it soon.

Just a question: what, if anything, remains Trac 18487 now? I.e. I see that you use "See" rather than "Closes" for it in the last commit message, does it mean that it's still not fixed even with these changes or am I just reading too much into it?

discnl commented:
Thanks for merging, and sorry about the title of the PR which only covered part of the changes (and apparently not the more discussed one).

Ticket 18487 is not completely or even adequately fixed, I left a [comment](http://trac.wxwidgets.org/ticket/18487#comment:17).
